src/emitters/spot.cpp

```diff
@@ -0,0 +1,176 @@
+        SurfaceInteraction3f si = zero<SurfaceInteraction3f>();
+        si.wavelengths = wavelengths;
+        Spectrum result = m_intensity->eval(si, active);
+
+        const Transform4f &trafo = m_world_transform->eval(time, active);
+        Vector3f local_dir = trafo.inverse() * -d;
+        local_dir /= norm(local_dir);
+        const Float cos_theta = local_dir.z();
+
+        si.uv = Point2f(0.5f + 0.5f * local_dir.x() / (local_dir.z() * m_uv_factor), 0.5f + 0.5f * local_dir.y() / (local_dir.z() * m_uv_factor));
+        result *= m_texture->eval(si, active);
+        
+        return select(cos_theta <= m_cos_cutoff_angle, Spectrum(0.0f), select(cos_theta >= m_cos_beam_width, result, result * ((m_cutoff_angle - acos(cos_theta)) * m_inv_transition_width)));
+    }
+
+    std::pair<Ray3f, Spectrum> sample_ray(Float time, Float wavelength_sample,
+                                          const Point2f &spatial_sample,
+                                          const Point2f & /*dir_sample*/,
+                                          Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::EndpointSampleRay, active);
+
+        // 1. Sample spectrum
+        auto [wavelengths, spec_weight] = sample_wavelength<Float, Spectrum>(wavelength_sample);
+
+        // 2. Sample directional component
+        const Transform4f &trafo = m_world_transform->eval(time, active);
+        Vector3f d = trafo * warp::square_to_uniform_cone(spatial_sample, m_cos_cutoff_angle);
+
+        return { Ray3f(trafo * Point3f(0.f), d, time, wavelengths),
+                select(m_beam_width == m_cutoff_angle,
+                        unpolarized<Spectrum>(spec_weight) * 2 * math::Pi<Float> * (1 - m_cos_cutoff_angle),
+                        unpolarized<Spectrum>(spec_weight) * (2 * math::Pi<Float> * (1 - (sin(m_cutoff_angle) - sin(m_beam_width))/(m_cutoff_angle - m_beam_width)))) };
+    }
+
+    std::pair<DirectionSample3f, Spectrum> sample_direction(const Interaction3f &it,
+                                                            const Point2f & /*sample*/,
+                                                            Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::EndpointSampleDirection, active);
+
+        auto trafo = m_world_transform->eval(it.time, active);
+
+        DirectionSample3f ds;
+        ds.p     = trafo.translation();
+        ds.n     = 0.f;
+        ds.uv    = 0.f;
+        ds.time  = it.time;
+        ds.pdf   = 1.f;
+        ds.delta = true;
+        ds.object = this;
+        ds.d     = (ds.p - it.p);
+        ds.dist  = norm(ds.d);
+        Float inv_dist = rcp(ds.dist);
+        ds.d *= inv_dist;
+
+        Spectrum spec = falloff_curve(ds.d, it.time, it.wavelengths, active) * (inv_dist * inv_dist);
+        return { ds, unpolarized<Spectrum>(spec) };
+    }
+
+    Float pdf_direction(const Interaction3f &, const DirectionSample3f &, Mask) const override {
+        return 0.f;
```

Comment:
I guess here it should return warp::squareToUniformConePdf(m_cosCutoffAngle) instead?

Reply:
Should I change line 125 to
`ds.pdf   = warp::square_to_uniform_cone_pdf(warp::square_to_uniform_cone(sample, m_cos_cutoff_angle), m_cos_cutoff_angle);`

and then at line 138, return ds.pdf?

Reply:
You will need to assign the result of `square_to_uniform_cone(sample, m_cos_cutoff_angle)` to `ds.d` as well.

And `pdf_direction()` should recompute the PDF given `ds.d` as it might be sampled from another distribution.

Reply:
But I am not given any `sample` in `pdf_direction()`. How do I call get a sample for  calling `square_to_uniform_cone_pdf()`?

Reply:
`square_to_uniform_cone_pdf` takes a 3D vector (the direction) not a 2D point on the unit square. You have this direction in `ds.d`. Also, you might have to transform it using the `m_world_transform` matrix.